docs/DesignDocs/libc_pthread_proposal.md

```diff
@@ -0,0 +1,80 @@
+* Child thread(Host): execute ECALL - OE_ECALL_THREAD_CREATE_ROUTINE if new thread is created successfully.
+* Child thread(Enclave): get the user defined hook function and execute it.
+
+**NOTE**: Currently the parameter of attributes for the created thread is unused. So the created threads are
+all joinable.
+
+### oe_thread_exit
+The function is used to exit the start_routine function.
+
+oe_thread_exit is an internal function which will be wrapped to an external API pthread_exit.
+
+The work sequence should be like,
+* Parent thread(Enclave): call pthread_create() to create a new thread and then return.
+* Child thread(Host): execute ECALL - OE_ECALL_THREAD_CREATE_ROUTINE.
+* Child thread(Enclave): call oe_setjmp() to store current execution context.
+* Child thread(Enclave): execute the user defined hook function.
+* Child thread(Enclave): call pthread_exit() inside user defined hook function.
+* Child thread(Enclave): in pthread_exit() call oe_longjmp() to restore the context.
+
+### oe_thread_join
+This function is used to wait for the specified thread to terminate.
+
+oe_thread_join is an internal function which will be wrapped to an external API pthread_join.
+
+The work sequence should be like,
+* Parent thread(Enclave): call pthread_create() to create a new thread and then return.
+* Other threads(Enclave/Host): call pthread_join() and then make an OCALL - OE_OCALL_THREAD_WAIT to wait for an event.
+* Child thread(Host): execute ECALL - OE_ECALL_THREAD_CREATE_ROUTINE.
+* Child thread(Host): wake up the waiting thread after OE_ECALL_THREAD_CREATE_ROUTINE return.
+* Other threads(Host): OE_OCALL_THREAD_WAIT return after being waked up.
+* Other threads(Enclave): in pthread_join() check the child thread status which is controlled in enclave before return.
+
+### oe_thread_self
+The function is used to get the calling thread’s ID. 
+
+oe_thread_self is an internal function which will be wrapped to an external API pthread_self.
+
+Currently it returns the td pointer. It will be changed to return an internal structure pointer which 
+is used in oe_thread_exit and oe_thread_join.
+
+```
+typedef struct _oe_thread_impl
+{
+    volatile oe_spinlock_t lock;
+    volatile oe_thread_state_t state;
+    oe_sgx_td_t* joiner_td;
+    oe_sgx_td_t* creater_td;
+    struct _oe_thread_common_queue_elm_t common_queue_elm;
+    void* (*start_routine)(void *);
+    void* arg;
+    void* retval;  // Used by oe_thread_exit() and oe_thread_join() to pass the exit message
+    void* tid;
+} oe_thread_impl_t
+```
+
+## Support OSes
+To get better performance, the OS thread created in host should be detached. Currently the OS thread can be created 
+by oe_thread_create which is a wrapper of pthread_create in Linux and CreateThread in Windows. The attribute parameter 
+is not used, so the detached thread can't be created with default attribute. However, if only Linux is supported, there will be 
+no issue.
```

Comment:
It should work on Windows too.  https://sourceforge.net/projects/pthreads4w/ for example implements it on windows so it's certainly possible.  https://stackoverflow.com/questions/12744324/how-to-detach-a-thread-on-windows-c discusses how to create a detached thread on Windows.

Reply:
OK. I will do a poc for this.

Reply:
Just checked https://sourceforge.net/projects/pthreads4w/, it is kind of complicated to implement the pthread_create. I think it is not suitable to include it in this patch.